### DISCUSSION TITLE
Hardlock after hitting Esc -> ALT+F4 during gameplay

### DISCUSSION THREAD

**Discussion by mty22** [ppy/osu#17209] [Q&A]
### Description:

The game will hardlock for a short period of time after hitting Esc (during gameplay) -> ALT + F4 combination relatively quickly.

I've been able replicate it several times (3 out of 5 attempts). Outside of replicating, I almost always ALT+F4 out of the game at random. I've had about 5 crashes over 2 months and only really looked into it more prior to writing this.

### How to replicate:

1. Load a beatmap.
2. Play for a short period of time (5-10s is fine), or fail the map.
3. Hit Esc -> ALT+F4 combo within a second of the failed or pause screen loading.

If successful, the game will hard lock for a reasonable amount of time before exiting to the desktop and will look like this:

![hardlock-screenshot](https://user-images.githubusercontent.com/2951231/157641848-3be34ad4-9ba8-4a58-91ed-f4bedfb782ea.jpg)

### Additional:
**osu!lazer version**: 2022.306.0.0
**OS**: Windows 10 (version: 21H1 Build: 19043.1586)
**Nvidia drivers**: 497.09

### Files:
[eventviewer_logs.zip](https://github.com/ppy/osu/files/8222504/eventviewer_logs.zip)
[runtime.log](https://github.com/ppy/osu/files/8222509/runtime.log)
[network.log](https://github.com/ppy/osu/files/8222510/network.log)
[performance.log](https://github.com/ppy/osu/files/8222511/performance.log)
[performance-draw.log](https://github.com/ppy/osu/files/8222512/performance-draw.log)
[performance-update.log](https://github.com/ppy/osu/files/8222513/performance-update.log)

Thanks.
frzn (osu!) / frznmatt (twitter/twitch).

**Comment by bdach** [comment]
for future reference please create discussions rather than issues directly. we will automatically escalate if it is needed.

event viewer log shows the following:

```
CoreCLR Version: 6.0.222.6406
.NET Version: 6.0.2
Description: The process was terminated due to an unhandled exception.
Exception Info: System.InvalidOperationException: Failed to acquire GL context (Video subsystem has not been initialized)
   at osu.Framework.Platform.SDL2.SDL2GraphicsBackend.MakeCurrent(IntPtr context)
   at osu.Framework.Threading.GameThread.processFrame()
   at osu.Framework.Threading.GameThread.RunSingleFrame()
   at osu.Framework.Threading.GameThread.&lt;createThread&gt;g__runWork|64_0()
   at System.Threading.Thread.StartHelper.Callback(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Thread.StartCallback()
```

i have zero idea how this can happen. possibly something to report upstream to sdl2, haven't looked much further than just examining the trace. if this is something we need to handle then this should be a framework issue anyhow.

**Reply by peppy** [reply]
My guess is that the Alt+F4 is triggering the window-level (SDL) cleanup sequence immediately, before our thread runners stop. Since we're pumping the draw thread using our own threading mechanism this could still be running at that point in time.
